### PR TITLE
[k8s_cloud_beta1] Sky down for a cluster deployed in Kubernetes to possibly remove sshjump pod.

### PR DESCRIPTION
**Pull request by aviweit**
Minimize the chances for having a dangling sshjump pod by adding a logic to `terminate_node` of Kubernetes provider - to remove sshjump pod and its service, in case sshjump pod is being reported as having its container not ready.

Otherwise, the sshjump pod is being reported with its main container ready where its lcm takes care of the cleaning i.e. terminates that pod and service.

### CODE REVIEW

**Comment by aviweit** [comment]
Hi @romilbhardwaj , I have added this capability per your comment in sshjump PR, where we may end with a dangling sshjump pod.

It seems to work fine in my environment. If you think it can help, please feel free to merge it.
Thanks.

File: sky/skylet/providers/kubernetes/node_provider.py [hunk 1]
```diff
@@ -240,6 +240,54 @@ def create_node(self, node_config, tags, count):
...
+                    sshjump_name, self.namespace)
+                containersReadyCondition = find(
+                    sshjump_pod.status.conditions, lambda c: c.type == 'ContainersReady')
+                if containersReadyCondition and \
+                    containersReadyCondition.status == 'False':
+                    # The main container is not ready. To be on the safe-side
+                    # and prevent a dangling sshjump pod - lets remove it and
+                    # the service. Otherwise main container is ready and its lcm
+                    # takes care of the cleaning
+                    kubernetes.core_api().delete_namespaced_pod(
```
Comment (romilbhardwaj) [inline sky/skylet/providers/kubernetes/node_provider.py:272]: Thanks @aviweit - a concern I have is this is splintering the ssh jump pod management code across modules. Jump pod creation is handled in `authentication.py::setup_kubernetes_authentication`, whereas this code is present in the node provider. Can we move the creation code also to node_provider?

Comment (romilbhardwaj) [reply sky/skylet/providers/kubernetes/node_provider.py:272]: Update - I have done some refactoring in `k8s_cloud_beta1` with the above suggestion. Can you merge the latest changes into this branch?

**Comment by aviweit** [comment]
@romilbhardwaj , I merged with `k8s_cloud_beta1` and moved the logic to `kubernetes_utils.py`.

It seems to work well in my environment.

Please find below tests I did:
1. launch task, wait for it to complete, remove cluster -> ssh jump LCM removed service and itself
2. deliberately injected wrong image name, launch task, wait for task to complete, did not finish, ^C, remove cluster -> cluster pod, services and ssh jump pod, service removed
3. launch task, wait for it to complete, deliberate remove ssh jump role, wait for ssh jump pod to fail, remove cluster -> cluster pod, services and ssh jump pod, service removed

File: sky/utils/kubernetes_utils.py [hunk 2]
```diff
@@ -333,6 +334,53 @@ def setup_sshjump_pod(sshjump_name: str, sshjump_image: str,
...
+        namespace: Namespace to remove the SSH jump pod and service from
+    """
+
+    def find(l, predicate):
+        """Utility function to find element in given list
+        """
+        results = [x for x in l if predicate(x)]
+        return results[0] if len(results) > 0 else None
+
+    sshjump_name = clouds.Kubernetes.SKY_SSH_JUMP_NAME
```
[the remark this answers is not in the document]
Comment (aviweit) [reply sky/utils/kubernetes_utils.py]: I pushed a change to retrieve sshjump pod name from the ray pod label rather than from `clouds.Kubernetes`.

**Review by romilbhardwaj — APPROVED** [review]
This looks great, thanks for the contribution @aviweit!

**Pull request merged**